src/createVue2SFCModule.ts

```diff
@@ -0,0 +1,203 @@
+import {
+	compileStyleAsync as sfc_compileStyleAsync,
+	compileTemplate as sfc_compileTemplate,
+	parse as sfc_parse,
+} from '@vue/component-compiler-utils'
+
+import * as vueTemplateCompiler from 'vue-template-compiler'
+import { VueTemplateCompiler } from '@vue/component-compiler-utils/dist/types'
+import { TemplateCompileOptions } from '@vue/component-compiler-utils/dist/compileTemplate'
+
+
+import {
+	parse as babel_parse,
+	ParserPlugin as babel_ParserPlugin
+} from '@babel/parser';
+
+import {
+	transformFromAstAsync as babel_transformFromAstAsync,
+	types as t,
+} from '@babel/core';
+
+// @ts-ignore (Could not find a declaration file for module '@babel/plugin-transform-modules-commonjs')
+import babelPluginTransformModulesCommonjs from '@babel/plugin-transform-modules-commonjs'
+
+
+import { formatError, withCache, hash, renameDynamicImport, parseDeps, interopRequireDefault, transformJSCode, loadDeps, createModule } from './tools.ts'
+
+import { Options, LoadModule, ModuleExport, CustomBlockCallback } from './types.ts'
+
+/**
+ * the version of the library (process.env.VERSION is set by webpack, at compile-time)
+ */
+const version : string = process.env.VERSION;
+
+const genSourcemap : boolean = !!process.env.GEN_SOURCEMAP;
+
+/**
+ * @internal
+ */
+const isProd : boolean = process.env.NODE_ENV === 'production';
+
+
+
+/**
+ * @internal
+ */
+
+export async function createSFCModule(source : string, filename : string, options : Options, loadModule : LoadModule) : Promise<ModuleExport> {
+
+	const component = {};
+
+
+	const { compiledCache, addStyle, log, additionalBabelPlugins = [], customBlockHandler } = options;
+
+	// vue-loader next: https://github.com/vuejs/vue-loader/blob/next/src/index.ts#L91
+	const descriptor = sfc_parse({
+		source,
+		filename,
+		needMap: genSourcemap,
+		compiler: vueTemplateCompiler as VueTemplateCompiler}
+		);
+
+	const customBlockCallbacks : CustomBlockCallback[] = customBlockHandler !== undefined ? await Promise.all( descriptor.customBlocks.map((block ) => customBlockHandler(block, filename, options)) ) : [];
+
+	const componentHash = hash(filename, version);
+	const scopeId = `data-v-${componentHash}`;
+
+	// hack: asynchronously preloads the language processor before it is required by the synchronous preprocessCustomRequire() callback, see below
+	if ( descriptor.template && descriptor.template.lang )
+		await loadModule(descriptor.template.lang, options);
+
+
+	const hasScoped = descriptor.styles.some(e => e.scoped);
+
+	const compileTemplateOptions : TemplateCompileOptions = descriptor.template ? {
+		// hack, since sourceMap is not configurable an we want to get rid of source-map dependency. see genSourcemap
+		source: descriptor.template.content,
+		filename,
+		compiler: vueTemplateCompiler as VueTemplateCompiler,
+		compilerOptions: undefined,
+		preprocessLang: descriptor.template.lang,
+		isProduction: isProd,
+		prettify: false
+	} : null;
```

Comment:
no scopeId here, should we handle scope attribute manually ?

Reply:
Yes, this is still a thing that doesn't work with vue2. It seems `vue-template-compiler` doesn't support the addition of the scope id data attribute on template root element. We need to find a way to add it manually to the component element. But it's not that simple as there's no reference to the componentDOM element inside `getStyle` callback. Maybe we should hack the generated `render` function ?